### PR TITLE
Add warnings if training args differ from checkpoint trainer state

### LINKED ISSUE #28867: Training arguments are not applied when resuming from a checkpoint
### System Info

I noticed when resuming the training of a model from a checkpoint changing properties like `save_steps` and `per_device_train_batch_size` has no effect. I'm wondering if there's something syntactically wrong here or technically the config of the model checkpoint overrides everything? I've seen a thread started about this [here](https://discuss.huggingface.co/t/if-train-resume-from-checkpoint-cant-change-trainerarguments/70715/4)

```
import transformers
from datetime import datetime

tokenizer.pad_token = tokenizer.eos_token

learning_rate = 5e-5  
warmup_steps = 100

gradient_accumulation_steps = 2  

trainer = transformers.Trainer(
    model=model,
    callbacks=[upload_checkpoint_callback],
    train_dataset=tokenized_train_dataset,
    eval_dataset=tokenized_val_dataset,
    args=transformers.TrainingArguments(
        output_dir=output_dir,
        warmup_steps=warmup_steps,
        per_device_train_batch_size=8,
        gradient_checkpointing=True,
        gradient_accumulation_steps=gradient_accumulation_steps,
        max_steps=5000,
        learning_rate=learning_rate,
        logging_steps=10,
        fp16=True,
        optim="paged_adamw_8bit",
        logging_dir="/content/logs",       
        save_strategy="steps",      
        save_steps=10,              
        evaluation_strategy="steps", 
        eval_steps=10,               
        load_best_model_at_end=True,
        report_to="wandb",           
        run_name=f"{run_name}-{datetime.now().strftime('%Y-%m-%d-%H-%M')}"          # Name of the W&B run (optional)
    ),
    data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False),
)

model.config.use_cache = False
trainer.train(resume_from_checkpoint="/content/latest_checkpoint/")
```

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Train a model from a checkpoint and adjust some training arguments, these won't have any effect

### Expected behavior

The training arguments should have effect when training off a checkpointed model

### PR DESCRIPTION
**Pull request by jonflynng**
# What does this PR do?

Resolves [#28867](https://github.com/huggingface/transformers/issues/28867)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
https://github.com/huggingface/transformers/issues/28867
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@ArthurZucker 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1370,6 +1370,29 @@ def ipex_optimize_model(self, model, training=False, dtype=torch.float32):
 
         return model
 
+    def compare_trainer_and_checkpoint_args(self, training_args, trainer_state):
+        attributes_map = {
```
Comment (jonflynng) [inline src/transformers/trainer.py:1374]: These are some parameters I expected to be overridden via the `training_args` when resuming from a checkpoint.

**Comment by ArthurZucker** [comment]
cc @muellerzr !

**Review by ArthurZucker — APPROVED** [review]
This LGTM, @muellerzr fine with you ?

**Review by muellerzr — APPROVED** [review]
Thanks! LG2M as well 🤗

**Pull request merged**